From 0f1808a6a89c71bf7dea1d351c9cd81007638a1a Mon Sep 17 00:00:00 2001
From: Kyrylo Yatsenko <hedrok@gmail.com>
Date: Thu, 26 Feb 2026 14:38:17 +0200
Subject: [PATCH] ospf6d: Fix router LSA when no new BDR elected

When on broadcast network only two routers are eligible to become
Designated Routers, one of them becomes DR (R1),
other one becomes BDR (R2).

If after that R1 goes down, R2 becomes DR and network has no BDR.

In this case on all other routers when R1 went down
`ospf6_router_lsa_originate` originates router LSA without this network
description as it has no fully adjacent Designated Router yet (it is not
elected).

But when R2 is elected as DR, no new router LSA is originated in
`neighbor_change`: it calls `ospf6_interface_state_change`, but state of
interface is unchanged, so it just returns -1. And for all other routers
network doesn't exist.

Fix this by originating router LSA in `neighbor_change` for this
specific case.

Signed-off-by: Kyrylo Yatsenko <hedrok@gmail.com>
---
 ospf6d/ospf6_interface.c | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/ospf6d/ospf6_interface.c b/ospf6d/ospf6_interface.c
index 7f16a96d4d..74aa013bfb 100644
--- a/ospf6d/ospf6_interface.c
+++ b/ospf6d/ospf6_interface.c
@@ -951,6 +951,7 @@ void backup_seen(struct event *event)
 void neighbor_change(struct event *event)
 {
 	struct ospf6_interface *oi;
+	int ret;
 
 	oi = (struct ospf6_interface *)EVENT_ARG(event);
 	assert(oi && oi->interface);
@@ -959,9 +960,15 @@ void neighbor_change(struct event *event)
 		zlog_debug("Interface Event %s: [NeighborChange]",
 			   oi->interface->name);
 
-	if (oi->state == OSPF6_INTERFACE_DROTHER ||
-	    oi->state == OSPF6_INTERFACE_BDR || oi->state == OSPF6_INTERFACE_DR)
-		ospf6_interface_state_change(dr_election(oi), oi);
+	if (oi->state == OSPF6_INTERFACE_DROTHER || oi->state == OSPF6_INTERFACE_BDR ||
+	    oi->state == OSPF6_INTERFACE_DR) {
+		ret = ospf6_interface_state_change(dr_election(oi), oi);
+
+		/* In this case router LSA is not originated elsewhere */
+		if (oi->area && oi->type == OSPF_IFTYPE_BROADCAST && ret == -1 &&
+		    oi->bdrouter == htonl(0))
+			OSPF6_ROUTER_LSA_SCHEDULE(oi->area);
+	}
 }
 
 void interface_down(struct event *event)
-- 
2.51.2

